src/speculators/data_generation/preprocessing.py

```diff
@@ -271,10 +316,14 @@ def build_eagle3_dataset(
         turn_dropout: If True, randomly keeps first N consecutive turns per
                      conversation
     """
-    # Detect or use provided assistant message pattern
-    if assistant_pattern is None:
+    # Detect and use provided assistant message pattern
+    supports_mask = _supports_assistant_mask(tokenizer)
+
+    if assistant_pattern is None and not supports_mask:
         assistant_pattern = _detect_assistant_pattern(tokenizer)
         log.info(f"Detected assistant pattern: {assistant_pattern[:80]}...")
+    elif supports_mask:
+        log.info("Using HF assistant token mask for loss masking")
     else:
         log.info(f"Using custom assistant pattern: {assistant_pattern[:80]}...")
```

Comment:
I think it would be better if we have the following flow (pseudocode):

```python
if assistant_pattern is not None:
    # Use assistant pattern
elif _supports_assistant_mask(tokenizer):
    # Use tokenizer w/ `return_assistant_tokens_mask=True`
else:
    assistant_pattern = _detect_assistant_pattern(tokenizer)
```

This way users can still specify an `assistant_pattern` to override the `return_assistant_tokens_mask` logic if they choose to.


The other change required for this is to make `_preprocess_batch` accept `str | None` for the assistant pattern and then we can assume we're using the tokenizer's automatic detection if the assistant pattern is `None`.